### PR TITLE
chore(eslint-plugin): enhance many rule descriptions

### LINKED ISSUE #4861: Docs: Proofread rule docs for clarity
### Suggested Changes

I've noticed many rules docs pages are unclear. We should go through and audit them for clarity.

For example, https://typescript-eslint.io/rules/member-ordering/ starts off rather unclear and doesn't explicitly say what `MemberType` is.

### Affected URL(s)

https://typescript-eslint.io/rules/*

### LINKED ISSUE #5064: Docs: Clarify & expand on explanation in no-for-in-array rule docs page
### Before You File a Documentation Request Please Confirm You Have Done The Following...

- [X] I have looked for existing [open or closed documentation requests](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3Adocumentation) that match my proposal.
- [X] I have [read the FAQ](https://typescript-eslint.io/docs/linting/troubleshooting) and my problem is not listed.

### Suggested Changes

Splitting out of #4861: https://typescript-eslint.io/rules/no-for-in-array still uses a copy&paste of the old TSLint docs. We should make the explanation more clear and link to relevant external resources (e.g. MDN) where useful.

### Affected URL(s)

https://typescript-eslint.io/rules/no-for-in-array

### PR DESCRIPTION
**Pull request by JoshuaKGoldberg**
## PR Checklist

- [x] Addresses an existing open issue: fixes #4861; fixes #5064
- [x] That issue was marked as [accepting prs](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aopen+is%3Aissue+label%3A%22accepting+prs%22)
- [x] Steps in [CONTRIBUTING.md](https://github.com/typescript-eslint/typescript-eslint/blob/main/CONTRIBUTING.md) were taken

## Overview

Goes through the ~90 core (non-extension, non-formatting) rules and spruces up the docs page. For each rule I tried to keep its documentation to roughly the following format:

1. A brief description of the TypeScript syntax at play, if relevant
2. What issue the rule is meant to tackle
3. How the rule tackles the issue

> This isn't 100% consistently kept to because I ran out of energy repeatedly throughout writing this horrendously boring PR. If you point out discrepancies that will help me regain the energy to fix them.

Simplifies the typical rule docs page format by joining the initial description with the _Rule Details_ section, leaving an _Examples_ section for just examples.

Pairs nicely with #5702.

### CODE REVIEW

File: .github/ISSUE_TEMPLATE/04-enhancement-new-base-rule-extension.yaml [hunk 1]
```diff
@@ -11,8 +11,9 @@ body:
     attributes:
       value: |
         # What is a "Base Rule Extension"?
-        In some cases, ESLint provides a rule itself, but it doesn't support TypeScript syntax; either it crashes, or it ignores the syntax, or it falsely reports against it.
-        In these cases, we create what we call an extension rule; a rule within our plugin that has the same functionality, but also supports TypeScript.
+        In some cases, ESLint provides a rule itself, but it doesn't support TypeScript syntax.
+        It may crash, or it ignore the syntax, or falsely reports against it.
+        In these cases we create what we call an extension rule: a rule within our plugin that has the same functionality, but also supports TypeScript.
```
Comment (JoshuaKGoldberg) [inline .github/ISSUE_TEMPLATE/04-enhancement-new-base-rule-extension.yaml:16]: Todo: split this copy change out

File: packages/website/src/theme/MDXComponents/RuleAttributes.tsx [hunk 2]
```diff
@@ -63,7 +48,7 @@ export function RuleAttributes({ name }: { name: string }): JSX.Element | null {
           <input
             type="checkbox"
             disabled
-            checked={!!rule.docs?.requiresTypeChecking}
+            checked={!!rule.docs.requiresTypeChecking}
```
Comment (JoshuaKGoldberg) [inline packages/website/src/theme/MDXComponents/RuleAttributes.tsx:51]: Todo: split this change out

File: packages/website/src/css/custom.css [hunk 3]
```diff
@@ -62,6 +62,8 @@ html[data-theme='dark']:root {
 
   --docsearch-muted-color: #aaa;
   --docusaurus-highlighted-code-line-bg: rgba(0, 0, 0, 0.3);
+
+  --gray-border-shadow: var(--ifm-color-secondary-dark);
```
Comment (JoshuaKGoldberg) [inline packages/website/src/css/custom.css:66]: Todo: split this change out

File: packages/eslint-plugin/src/rules/comma-spacing.ts [hunk 4]
```diff
@@ -19,7 +19,7 @@ type MessageIds = 'unexpected' | 'missing';
 export default createRule<Options, MessageIds>({
   name: 'comma-spacing',
   meta: {
-    type: 'suggestion',
+    type: 'layout',
```
Comment (JoshuaKGoldberg) [inline packages/eslint-plugin/src/rules/comma-spacing.ts:22]: Todo: split this fix out

File: packages/website/sidebars/sidebar.rules.js [hunk 5]
```diff
@@ -13,14 +13,19 @@ const rules = Object.entries(plugin.rules).map(([name, rule]) => {
 const deprecatedRules = new Set(rules.filter(rule => rule.meta.deprecated));
 
 const formattingRules = new Set(
+  rules.filter(rule => !rule.meta.deprecated && rule.meta.type === 'layout'),
+);
```
Comment (JoshuaKGoldberg) [inline packages/website/sidebars/sidebar.rules.js:17]: Todo: split out

File: packages/website/src/theme/MDXComponents/Feature.module.css [hunk 6]
```diff
@@ -0,0 +1,22 @@
+.feature {
```
Comment (JoshuaKGoldberg) [inline packages/website/src/theme/MDXComponents/Feature.module.css:1]: Todo: split out

File: packages/website/docusaurusConfig.ts [hunk 7]
```diff
@@ -141,7 +141,7 @@ const themeConfig: ThemeCommonConfig & AlgoliaThemeConfig = {
       },
     ],
     // style: 'primary',
-    copyright: `Copyright © ${new Date().getFullYear()} TypeScript ESLint, Inc. Built with Docusaurus.`,
+    copyright: `Copyright © ${new Date().getFullYear()} TypeScript ESLint, Inc. Built with <a href="">Docusaurus</a>.`,
```
Comment (JoshuaKGoldberg) [inline packages/website/docusaurusConfig.ts:144]: Todo: remove, I was just experimenting...

File: packages/eslint-plugin/docs/rules/await-thenable.md [hunk 8]
```diff
@@ -6,12 +6,13 @@ description: 'Disallow awaiting a value that is not a Thenable.'
...
 > See **https://typescript-eslint.io/rules/await-thenable** for documentation.
 
-This rule disallows awaiting a value that is not a "Thenable" (an object which has `then` method, such as a Promise).
-While it is valid JavaScript to await a non-`Promise`-like value (it will resolve immediately), this pattern is often a programmer error, such as forgetting to add parenthesis to call a function that returns a Promise.
+A "Thenable" value is an object with has a `then` method, such as a Promise.
+The `await` keyword is generally used to retrieve the result of calling a Thenable's `then` method.
 
-## Rule Details
+If the `await` keyword is used on a value that is not a Thenable, the value is directly resolved immediately.
+While doing so is valid JavaScript, it i often a programmer error, such as forgetting to add parenthesis to call a function that returns a Promise.
```
[the remark this answers is not in the document]
Comment (JoshuaKGoldberg) [reply packages/eslint-plugin/docs/rules/await-thenable.md:13]: Aha, thanks 😄

File: packages/eslint-plugin/docs/rules/no-non-null-assertion.md [hunk 9]
```diff
@@ -6,36 +6,39 @@ description: 'Disallow non-null assertions using the `!` postfix operator.'
...
-## Rule Details
+TypeScript's `!` non-null assertion operator asserts to the type system that an expression is non-nullable, as in not `null` or `undefined`.
+Using assertions to tell the type system new information is often a sign that code is not fully type-safe.
+It's generally better to structure program logic so that TypeScript understands when values may be nullable.
 
-Using non-null assertions cancels the benefits of the strict null-checking mode.
+## Examples
 
-Examples of code for this rule:
+<!-- Using non-null assertions cancels the benefits of the strict null-checking mode. -->
```
Comment (bradzacher) [inline packages/eslint-plugin/docs/rules/no-non-null-assertion.md:15]: is this commented out intentionally?

**Review by bradzacher — DISMISSED** [review]
image

File: packages/eslint-plugin/docs/rules/no-non-null-assertion.md (hunk 9 above)
Comment (JoshuaKGoldberg) [reply packages/eslint-plugin/docs/rules/no-non-null-assertion.md]: ...huh, I have no recollection of commenting this.

Comment (bradzacher) [reply packages/eslint-plugin/docs/rules/no-non-null-assertion.md]: was it supposed to be deleted?
It's text within the "examples" section.

Comment (JoshuaKGoldberg) [reply packages/eslint-plugin/docs/rules/no-non-null-assertion.md]: Oh, thank you, yes - that

**Review by bradzacher — APPROVED** [review]

**Comment by JoshuaKGoldberg** [comment]
...well this is frustrating. GitHub Actions froze at `Waiting for a runner to pick up this job...`. I tried canceling the job and starting a new one. I think GHA is undergoing an incident.

I can run `yarn build`, `yarn lint`, and `yarn test` locally to verify the merging from `main` did not break the world.

**Pull request merged**